Use Sphinx for all documentation

ddn0 opened the pull request:
Previously, Python documentation (Sphinx) and C++ documentation
(Doxygen) lived parallel and non-intersecting lives.

Instead, use the breathe extension to convert Doxygen XML output into
Sphinx, and use Sphinx as the container for all documentation.

Disable non-XML Doxygen outputs for simplicity.

Apply a slightly more information-dense Sphinx template based on pydata
rather than using one of the Sphinx defaults, alabaster.

Because it doesn't make too much sense to have multiple documentation
entry points, move from documentation being per (CMake) target to being
per project, and correspondingly, move docs to $SRC_DIR/docs rather than
having documentation spread throughout the project directories.

Along the way, decouple documentation generation from setup.py.
Documentation generation now depends on a completed build and uses the
Python environment script to find its dependencies. This makes it more
like any other command that depends on the Katana library.

tylershunt commented:
Word. Thanks this looks good.

Tip about github if you use one of the keywords like 'closes' (https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords) github will close the issue for you when this PR merges.

ddn0 committed:
Use Sphinx for all documentation

Previously, Python documentation (Sphinx) and C++ documentation
(Doxygen) lived parallel and non-intersecting lives.

Instead, use the breathe extension to convert Doxygen XML output into
Sphinx, and use Sphinx as the container for all documentation.

Disable non-XML Doxygen outputs for simplicity.

Apply a slightly more information-dense Sphinx template based on pydata
rather than using one of the Sphinx defaults, alabaster.

Because it doesn't make too much sense to have multiple documentation
entry points, move from documentation being per (CMake) target to being
per project, and correspondingly, move docs to $SRC_DIR/docs rather than
having documentation spread throughout the project directories.

Along the way, decouple documentation generation from setup.py.
Documentation generation now depends on a completed build and uses the
Python environment script to find its dependencies. This makes it more
like any other command that depends on the Katana library.

arthurp commented on cmake/Modules/KatanaBuildCommon.cmake:
```diff
@@ -402,32 +407,51 @@ install(
+    return()
+  endif ()
+
+  set(PYTHON_ENV_SCRIPT ${CMAKE_CURRENT_BINARY_DIR}/python_env.sh)
```
I would encourage you to get this from the `PYTHON_ENV_SCRIPT` target property of the appropriate Python cmake target. Instead of just relying on this being called in the same directory as that target.

arthurp commented on docs/conf.py:
```diff
@@ -1,32 +1,27 @@
+
+# Breathe takes a minute or two to parse Doxygen output. If you aren't editing
+# C++ documents, set this to false for faster edit-render loops.
+cxx_enabled = True
```
Provide an environment variable? That way people can more easily set it in there build without risk of committing the change.

arthurp commented on docs/conf.py:
```diff
@@ -50,14 +51,27 @@
+author = "Katana Graph"
+
+# TODO(ddn): Get this from katana.libgalois.version
+copyright = "Katana Graph 2021"
```
Shouldn't this be "Katana Graph, Inc."?

arthurp commented on docs/katana.property_graph.rst:
```diff
@@ -1,5 +1,6 @@
-Property Graphs (katana.property_graph)
-=======================================
+===============
+Property Graphs
+===============
```
If we are standardizing on "above and below" heading style, can we have a checker for it at some point? (Also I personally slightly prefer the under only style, but I don't really care much.)

ddn0 replied:
I actually don't care about the style either but from my reading of

- https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections

over and underlining is a different section size than just underling, so I thought to use that for titles of pages so that the rest of the page has more section styles to use.

arthurp replied:
ooh. Thanks for mentioning this. I wasn't careful enough. I agree with your change.

arthurp commented:
One more thing: The generated docs appear to have TOC for the C++ documentation. This kinda makes them useless. Also all the C++ docs are one giant page. If they can be split into smaller pages that would be nice. Makes the scroll bar more useful.

I have no idea where these changes would be made since I don't understand breathe. Sorry.

ddn0 commented:
> One more thing: The generated docs appear to have TOC for the C++ documentation. This kinda makes them useless. Also all the C++ docs are one giant page. If they can be split into smaller pages that would be nice. Makes the scroll bar more useful.
>
> I have no idea where these changes would be made since I don't understand breathe. Sorry.

Yeah. The next step would be to replace `doxygenindex` https://breathe.readthedocs.io/en/latest/directives.html#doxygenindex with a set of pages that provide Sphinx directives to document specific symbols rather than everything under the sun.

arthurp commented:
>  The next step would be to replace `doxygenindex` https://breathe.readthedocs.io/en/latest/directives.html#doxygenindex with a set of pages that provide Sphinx directives to document specific symbols rather than everything under the sun.

I think this is something that should happen sooner rather than later (not this PR, but still). The C++ docs are unusable to me at the moment. So I would prioritize this or at least make sure to track it and keep it from slipping.

ddn0 committed:
Add sphinx-tabs
